From 291a94aef9c5671f8a33ccabc392d360b3e0df0a Mon Sep 17 00:00:00 2001
From: "yun.cai" <yun.cai@amlogic.com>
Date: Tue, 11 Mar 2014 16:56:19 +0800
Subject: [PATCH 3827/5965] PD#87078:update I2C driver,include 3 commits:

Revert "PD #86922: i2c distinguish M6(port AO A B) and M8(port AO A B C D)"

This reverts commit 25abc12a6d2fcc60316baf5ee830ece6d2d11734.

PD #87078: i2c: move device config properties to dtd

PD #86922: i2c: fix show_i2c_mode error when one i2c device is disabled
---
 arch/arm/boot/dts/amlogic/meson6_g18.dtd      |   6 +
 arch/arm/boot/dts/amlogic/meson6_g18_512M.dtd |   6 +
 arch/arm/boot/dts/amlogic/meson6_g33_1212.dtd |   7 +-
 arch/arm/boot/dts/amlogic/meson6_g33_1G.dtd   |   7 +-
 arch/arm/boot/dts/amlogic/meson6_g33_512M.dtd |   7 +-
 arch/arm/boot/dts/amlogic/meson6_g35.dtd      |   6 +
 arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd  |  10 ++
 .../arm/boot/dts/amlogic/meson8_k101_512M.dtd |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k150_v1.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k160_v1.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd  |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd  |  10 ++
 .../boot/dts/amlogic/meson8_k200a_2G_4k2k.dtd |  10 ++
 .../dts/amlogic/meson8_k200a_2G_hdmiin.dtd    |  10 ++
 arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd |   6 +
 arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd |   6 +
 .../boot/dts/amlogic/meson8_k200b_2G_4k2k.dtd |  10 ++
 .../dts/amlogic/meson8_k200b_2G_hdmiin.dtd    |  10 ++
 arch/arm/boot/dts/amlogic/meson8_pxp.dtd      |  10 ++
 arch/arm/boot/dts/amlogic/meson8_skt.dtd      |  10 ++
 arch/arm/boot/dts/amlogic/meson8b_skt.dtd     |  10 ++
 drivers/amlogic/i2c/aml_i2c.c                 | 170 +++++++-----------
 include/linux/i2c-aml.h                       |   1 +
 26 files changed, 270 insertions(+), 112 deletions(-)

diff --git a/arch/arm/boot/dts/amlogic/meson6_g18.dtd b/arch/arm/boot/dts/amlogic/meson6_g18.dtd
index b64bd0ba775b..451018f25055 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g18.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g18.dtd
@@ -425,6 +425,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -442,6 +444,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;	
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -459,6 +463,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson6_g18_512M.dtd b/arch/arm/boot/dts/amlogic/meson6_g18_512M.dtd
index b96ccc5f75c2..87da22a49fff 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g18_512M.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g18_512M.dtd
@@ -425,6 +425,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -442,6 +444,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;	
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -459,6 +463,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson6_g33_1212.dtd b/arch/arm/boot/dts/amlogic/meson6_g33_1212.dtd
index ec24362c54f4..7fc38f2e84fb 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g33_1212.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g33_1212.dtd
@@ -445,6 +445,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 	
 //$$ DEVICE = "I2C_A"
@@ -462,6 +464,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -479,7 +483,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
-
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson6_g33_1G.dtd b/arch/arm/boot/dts/amlogic/meson6_g33_1G.dtd
index 723b4074c580..2ba695f6c06e 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g33_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g33_1G.dtd
@@ -445,6 +445,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 	
 //$$ DEVICE = "I2C_A"
@@ -462,6 +464,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -479,7 +483,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
-
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson6_g33_512M.dtd b/arch/arm/boot/dts/amlogic/meson6_g33_512M.dtd
index 44abdfbd7f1c..1a2be0efafaf 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g33_512M.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g33_512M.dtd
@@ -460,6 +460,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 	
 //$$ DEVICE = "I2C_A"
@@ -477,6 +479,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -494,7 +498,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
-
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson6_g35.dtd b/arch/arm/boot/dts/amlogic/meson6_g35.dtd
index f91dc871e6f6..299cfe4a074c 100755
--- a/arch/arm/boot/dts/amlogic/meson6_g35.dtd
+++ b/arch/arm/boot/dts/amlogic/meson6_g35.dtd
@@ -457,6 +457,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -474,6 +476,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;	
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -491,6 +495,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
index 31d3d71b34e3..8368a94f3bbb 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_1G.dtd
@@ -535,6 +535,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -552,6 +554,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -569,6 +573,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -585,6 +591,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -602,6 +610,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd b/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
index f0105686b79b..f5d063357eb5 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k100_2G.dtd
@@ -545,6 +545,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -562,6 +564,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -579,6 +583,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -595,6 +601,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -612,6 +620,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
index 25c4f5eae2e8..12849355bb7d 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_1G.dtd
@@ -467,6 +467,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -484,6 +486,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -501,6 +505,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -517,6 +523,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -534,6 +542,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd b/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
index 87bd205b5c04..6826eec6d5a7 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k101_512M.dtd
@@ -468,6 +468,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -485,6 +487,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -502,6 +506,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -518,6 +524,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -535,6 +543,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
index ae14645dba61..17ec8375f355 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k102_v1.dtd
@@ -546,6 +546,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -563,6 +565,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -580,6 +584,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -596,6 +602,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -613,6 +621,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k150_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k150_v1.dtd
index bb551a5f4fbb..983d3b3f706e 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k150_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k150_v1.dtd
@@ -467,6 +467,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -484,6 +486,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -501,6 +505,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -517,6 +523,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -534,6 +542,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k160_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k160_v1.dtd
index e2c86d5e66fe..ab051324a610 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k160_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k160_v1.dtd
@@ -467,6 +467,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -484,6 +486,8 @@ sdio{
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -501,6 +505,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -517,6 +523,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -534,6 +542,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
index 0df6fb414f4f..a6a9a9ae6153 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_1G.dtd
@@ -620,6 +620,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -636,6 +638,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -653,6 +657,8 @@ sdio{
 //		pinctrl-0=<&b_i2c_master>;
 //		#address-cells = <1>;
 //		#size-cells = <0>;
+//    use_pio = <0>;
+//		master_i2c_speed = <300000>;
 // 	};
 
 //$$ DEVICE="I2C_C"
@@ -669,6 +675,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -686,6 +694,8 @@ sdio{
 //		pinctrl-0=<&d_i2c_master>;
 //		#address-cells = <1>;
 //		#size-cells = <0>;
+//    use_pio = <0>;
+//		master_i2c_speed = <300000>;
 //	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
index 3696bb5c9a7f..1e230d6ecd31 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200_v1.dtd
@@ -515,6 +515,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -531,6 +533,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -548,6 +552,8 @@ sdio{
 //		pinctrl-0=<&b_i2c_master>;
 //		#address-cells = <1>;
 //		#size-cells = <0>;
+//    use_pio = <0>;
+//		master_i2c_speed = <300000>;
 // 	};
 
 //$$ DEVICE="I2C_C"
@@ -564,6 +570,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -581,6 +589,8 @@ sdio{
 //		pinctrl-0=<&d_i2c_master>;
 //		#address-cells = <1>;
 //		#size-cells = <0>;
+//    use_pio = <0>;
+//		master_i2c_speed = <300000>;
 //	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200a_2G_4k2k.dtd b/arch/arm/boot/dts/amlogic/meson8_k200a_2G_4k2k.dtd
index ad40fcde40d0..1872d9c8179c 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200a_2G_4k2k.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200a_2G_4k2k.dtd
@@ -506,6 +506,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -522,6 +524,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -539,6 +543,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -555,6 +561,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -572,6 +580,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200a_2G_hdmiin.dtd b/arch/arm/boot/dts/amlogic/meson8_k200a_2G_hdmiin.dtd
index 1da1e2c15392..d1b65e69b8d9 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200a_2G_hdmiin.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200a_2G_hdmiin.dtd
@@ -524,6 +524,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -540,6 +542,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -557,6 +561,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -573,6 +579,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -590,6 +598,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
index 6e0929bb4533..2226506b8e97 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_1G.dtd
@@ -621,6 +621,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -637,6 +639,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -670,6 +674,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
index 94da72eaff6c..5b405f8b6cc0 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G.dtd
@@ -611,6 +611,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -627,6 +629,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -660,6 +664,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_4k2k.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_4k2k.dtd
index bdc1f4a1bd4a..6686288dc489 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_4k2k.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_4k2k.dtd
@@ -504,6 +504,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -520,6 +522,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -537,6 +541,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -553,6 +559,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -570,6 +578,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_hdmiin.dtd b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_hdmiin.dtd
index c91a536fa56c..2df0f88d0d48 100755
--- a/arch/arm/boot/dts/amlogic/meson8_k200b_2G_hdmiin.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_k200b_2G_hdmiin.dtd
@@ -524,6 +524,8 @@ sdio{
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -540,6 +542,8 @@ sdio{
 //		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -557,6 +561,8 @@ sdio{
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -573,6 +579,8 @@ sdio{
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -590,6 +598,8 @@ sdio{
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_pxp.dtd b/arch/arm/boot/dts/amlogic/meson8_pxp.dtd
index 91ea6798b0f5..6ddcddbb8607 100755
--- a/arch/arm/boot/dts/amlogic/meson8_pxp.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_pxp.dtd
@@ -426,6 +426,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -443,6 +445,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -460,6 +464,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -476,6 +482,8 @@ void root_func(){
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -493,6 +501,8 @@ void root_func(){
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8_skt.dtd b/arch/arm/boot/dts/amlogic/meson8_skt.dtd
index 48c31b86fe58..55d31fd70b26 100755
--- a/arch/arm/boot/dts/amlogic/meson8_skt.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8_skt.dtd
@@ -470,6 +470,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -487,6 +489,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -504,6 +508,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -520,6 +526,8 @@ void root_func(){
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -537,6 +545,8 @@ void root_func(){
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/arch/arm/boot/dts/amlogic/meson8b_skt.dtd b/arch/arm/boot/dts/amlogic/meson8b_skt.dtd
index 7b6aa5beec40..56728ab34a7c 100755
--- a/arch/arm/boot/dts/amlogic/meson8b_skt.dtd
+++ b/arch/arm/boot/dts/amlogic/meson8b_skt.dtd
@@ -476,6 +476,8 @@ void root_func(){
 		device_id = <0>;
 		pinctrl-names="default";
 		pinctrl-0=<&ao_i2c_master>;
+		use_pio = <0>;
+		master_i2c_speed = <100000>;
 	};
 
 //$$ DEVICE = "I2C_A"
@@ -493,6 +495,8 @@ void root_func(){
 		pinctrl-0=<&a_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 	
 //$$ DEVICE="I2C_B"
@@ -510,6 +514,8 @@ void root_func(){
 		pinctrl-0=<&b_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
  	};
 
 //$$ DEVICE="I2C_C"
@@ -526,6 +532,8 @@ void root_func(){
 //		pinctrl-0=<&c_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 //$$ DEVICE="I2C_D"
@@ -543,6 +551,8 @@ void root_func(){
 		pinctrl-0=<&d_i2c_master>;
 		#address-cells = <1>;
 		#size-cells = <0>;
+		use_pio = <0>;
+		master_i2c_speed = <300000>;
 	};
 
 /// ***************************************************************************************
diff --git a/drivers/amlogic/i2c/aml_i2c.c b/drivers/amlogic/i2c/aml_i2c.c
index e81af89e9f72..aaa2e74d9050 100755
--- a/drivers/amlogic/i2c/aml_i2c.c
+++ b/drivers/amlogic/i2c/aml_i2c.c
@@ -56,20 +56,10 @@ static irqreturn_t aml_i2c_complete_isr(int irq, void *dev_id)
 #include <linux/pinctrl/consumer.h>
 #include <linux/of_i2c.h>
 #include <linux/of_address.h>
-struct aml_i2c_property{
-	char name[24];
-	int id;
-	int index;
-	kernel_ulong_t drv_data;
-};
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
-#define AML_I2C_DEVICE_NUM		5
-#else
-#define AML_I2C_DEVICE_NUM		3
-#endif
-static struct aml_i2c_property aml_i2c_properties_config[];
-
 
+int aml_i2c_device_num = 0;
+static struct aml_i2c_platform *aml_i2c_properties_list;
+#define AML_I2C_PRINT(fmt,args...)	printk(KERN_DEBUG "[aml_i2c]" fmt,##args)
 
 static void aml_i2c_set_clk(struct aml_i2c *i2c, unsigned int speed)
 {
@@ -783,9 +773,17 @@ static ssize_t show_i2c_mode(struct class *class, struct class_attribute *attr,
     struct i2c_adapter *i2c_adap;
     struct aml_i2c *i2c;
     int i;
-    for (i=0; i<AML_I2C_DEVICE_NUM; i++) {
-      i2c_adap = i2c_get_adapter(i);
+    int device_id;
+    struct aml_i2c_platform * aml_i2c_property;
+    aml_i2c_property = aml_i2c_properties_list;
+
+    for (i=0; i<aml_i2c_device_num; i++) {
+		AML_I2C_PRINT("%s %s %d: i= %d\n",__FILE__,__func__,__LINE__,i);
+			device_id = aml_i2c_property->master_no;
+			aml_i2c_property =container_of((aml_i2c_property->list.next), struct aml_i2c_platform, list.next);
+      i2c_adap = i2c_get_adapter(device_id);
 	    i2c = i2c_get_adapdata(i2c_adap);
+
       printk("i2c(%d) work in mode %d\n", i, i2c->mode);
 	  }
     return 1;
@@ -817,7 +815,13 @@ static ssize_t store_i2c_mode(struct class *class, struct class_attribute *attr,
         i2c->aml_i2c_hrtimer.function = aml_i2c_hrtimer_notify;
       }
       else if (mode == I2C_INTERRUPT_MODE) {
-        plat = (struct aml_i2c_platform *)aml_i2c_properties_config[bus_num].drv_data;
+				struct list_head *p;
+				list_for_each(p, &aml_i2c_properties_list->list){
+					plat = list_entry(p,struct aml_i2c_platform, list);
+					if(plat->master_no == bus_num)
+						break;
+				}
+//        plat = (struct aml_i2c_platform *)aml_i2c_properties_config[bus_num].drv_data;
 			  i2c->irq = plat->use_pio >> 2;
         ret = request_irq(i2c->irq, aml_i2c_complete_isr, IRQF_DISABLED, "aml_i2c", i2c);
         printk("i2c master(%d) request irq(%d) %s\n", bus_num, i2c->irq, 
@@ -877,15 +881,53 @@ static int aml_i2c_probe(struct platform_device *pdev)
 			dev_err(&pdev->dev, "no platform data\n");
 			return -EINVAL;
 	}
-	
+
 	ret = of_property_read_u32(pdev->dev.of_node,"device_id",&device_id);
 	if(ret){
 			printk("don't find to match device_id\n");
 			return -1;
 	}
-		
+
 	pdev->id = device_id;
-	plat = (struct aml_i2c_platform*)aml_i2c_properties_config[device_id].drv_data;
+
+		struct aml_i2c_platform *aml_i2c_property = kzalloc(sizeof(struct aml_i2c_platform), GFP_KERNEL);
+
+		if(!aml_i2c_property)
+			printk("can't alloc mem for i2c_property\n");
+		else{
+			ret = of_property_read_u32(pdev->dev.of_node, "use_pio", &(aml_i2c_property->use_pio));
+			if(ret){
+				printk("not find match use_pio, use default\n"); //(INT_I2C_MASTER2<<2)|I2C_INTERRUPT_MODE,
+				aml_i2c_property->use_pio = 0;
+			}
+
+			ret = of_property_read_u32(pdev->dev.of_node, "master_i2c_speed", &(aml_i2c_property->master_i2c_speed));
+			if(ret){
+				printk("not find match master_i2c_speed, use default\n");
+				if(aml_i2c_device_num == 0)
+					aml_i2c_property->master_i2c_speed = 100000;
+				else
+					aml_i2c_property->master_i2c_speed = 300000;
+			}
+
+			aml_i2c_property->wait_count         	= 50000;
+			aml_i2c_property->wait_ack_interval 	= 5;
+			aml_i2c_property->wait_read_interval = 5;
+			aml_i2c_property->wait_xfer_interval = 5;
+			aml_i2c_property->master_no          = device_id;
+			aml_i2c_property->master_state_name  = NULL;
+		}
+
+		if(aml_i2c_device_num == 0){
+			aml_i2c_properties_list = aml_i2c_property;
+			INIT_LIST_HEAD(&aml_i2c_properties_list->list);
+		}
+		else{
+			list_add_tail(&aml_i2c_property->list ,&aml_i2c_properties_list->list);
+		}
+		aml_i2c_device_num++;
+
+	plat = (struct aml_i2c_platform*)aml_i2c_property;
 
 	ret=of_property_read_string(pdev->dev.of_node,"pinctrl-names",&plat->master_state_name);
 	printk("plat->state_name:%s\n",plat->master_state_name);
@@ -1026,96 +1068,6 @@ static int aml_i2c_remove(struct platform_device *pdev)
     return 0;
 }
 
-#ifdef CONFIG_OF
-
-//static bool pinmux_dummy_share(bool select)
-//{
-//    return select;
-//}
-
-static struct aml_i2c_platform aml_i2c_driver_data_ao = {
-    .wait_count         = 50000,
-    .wait_ack_interval  = 5,
-    .wait_read_interval = 5,
-    .wait_xfer_interval = 5,
-    .master_no          = AML_I2C_MASTER_AO,
-    .use_pio            = 0, //(INT_I2C_MASTER_AO<<2)|I2C_INTERRUPT_MODE,
-    .master_i2c_speed   = AML_I2C_SPPED_100K,
-    .master_state_name	= NULL,
-};
-
-static struct aml_i2c_platform aml_i2c_driver_data_a = {
-    .wait_count             = 50000,
-    .wait_ack_interval   = 5,
-    .wait_read_interval  = 5,
-    .wait_xfer_interval   = 5,
-    .master_no          = AML_I2C_MASTER_A,
-    .use_pio            = 0,//(INT_I2C_MASTER0<<2)|I2C_INTERRUPT_MODE,
-    .master_i2c_speed   = AML_I2C_SPPED_300K,
-    .master_state_name  = NULL,
-};
-
-static struct aml_i2c_platform aml_i2c_driver_data_b = {
-    .wait_count         = 50000,
-    .wait_ack_interval = 5,
-    .wait_read_interval = 5,
-    .wait_xfer_interval = 5,
-    .master_no          = AML_I2C_MASTER_B,
-    .use_pio            = 0,//(INT_I2C_MASTER1<<2)|I2C_INTERRUPT_MODE,
-    .master_i2c_speed   = AML_I2C_SPPED_300K,
-    .master_state_name  = NULL,
-};
-
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
-static struct aml_i2c_platform aml_i2c_driver_data_c = {
-    .wait_count         = 50000,
-    .wait_ack_interval = 5,
-    .wait_read_interval = 5,
-    .wait_xfer_interval = 5,
-    .master_no          = AML_I2C_MASTER_C,
-    .use_pio            = 0,//(INT_I2C_MASTER2<<2)|I2C_INTERRUPT_MODE,
-    .master_i2c_speed   = AML_I2C_SPPED_300K,
-    .master_state_name  = NULL,
-};
-
-static struct aml_i2c_platform aml_i2c_driver_data_d = {
-    .wait_count         = 50000,
-    .wait_ack_interval = 5,
-    .wait_read_interval = 5,
-    .wait_xfer_interval = 5,
-    .master_no          = AML_I2C_MASTER_D,
-    .use_pio            = 0,//(INT_I2C_MASTER3<<2)|I2C_INTERRUPT_MODE,
-    .master_i2c_speed   = AML_I2C_SPPED_300K,
-    .master_state_name  = NULL,
-};
-#endif
-
-static struct aml_i2c_property aml_i2c_properties_config[AML_I2C_DEVICE_NUM]={
-	{
-		.name = "device_id",
-		.drv_data = ((kernel_ulong_t)&aml_i2c_driver_data_ao),
-	},
-	{
-		.name = "device_id",
-		.drv_data = ((kernel_ulong_t)&aml_i2c_driver_data_a),
-	},
-	{
-		.name = "device_id",
-		.drv_data = ((kernel_ulong_t)&aml_i2c_driver_data_b),
-	},
-#if MESON_CPU_TYPE >= MESON_CPU_TYPE_MESON8
-	{
-		.name = "device_id",
-		.drv_data = ((kernel_ulong_t)&aml_i2c_driver_data_c),
-	},
-	{
-		.name = "device_id",
-		.drv_data = ((kernel_ulong_t)&aml_i2c_driver_data_d),
-	},
-#endif
-};
-
-#endif
 
 #ifdef CONFIG_OF
 static const struct of_device_id meson6_i2c_dt_match[]={
diff --git a/include/linux/i2c-aml.h b/include/linux/i2c-aml.h
index 27091d83063b..1ddf267f8151 100755
--- a/include/linux/i2c-aml.h
+++ b/include/linux/i2c-aml.h
@@ -101,6 +101,7 @@ struct aml_i2c_platform{
 	pinmux_set_t  master_a_pinmux;
 	pinmux_set_t  master_b_pinmux;
 	const char *master_state_name;
+	struct list_head list;
 };
 
 /**************i2c software gpio***************/
-- 
2.19.0

